### PR TITLE
sof: drc: fix drc_inv_fixed function and refine comments

### PR DESCRIPTION
**Pull request by johnylin76**
Used ABS(x) instead of x for comparing to ONE_OVERT_SQRT2(~0.707)
because x could be a negative value and here we want to transfrom x
from range (-1, -0.5];[0.5, 1) to (-1, -0.707];[0.707, 1)

Revised code line 237~241 to be easier to understand.

Refined some comments to prevent confusion of function usage.

### CODE REVIEW

File: src/audio/drc/drc_math_generic.c [hunk 1]
```diff
@@ -118,7 +118,7 @@ inline int32_t drc_sin_fixed(int32_t x)
 }
```
Comment (dbaluta) [inline src/audio/drc/drc_math_generic.c:119]: Your commit is doing two things when it should do only one! 

Please move comments updates on a separate patch.

**Review by ShriramShastry — COMMENTED** [review]
Changes looks good!
int precision_inv; change to int32_t precision_inv;
 change qc to QC

File: src/audio/drc/drc_math_generic.c [hunk 2]
```diff
@@ -207,14 +211,15 @@ inline int32_t drc_inv_fixed(int32_t x, int precision_x, int precision_y)
 	const int32_t A1 = Q_CONVERT_FLOAT(-21.25031280517578125f, qc);
 	const int32_t A0 = Q_CONVERT_FLOAT(7.152250766754150390625f, qc);
 	int e;
```
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:213]: Change int e to int_32_t

Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:213]: Change int e to int_32_t

File: src/audio/drc/drc_math_generic.c [hunk 3]
```diff
@@ -207,14 +211,15 @@ inline int32_t drc_inv_fixed(int32_t x, int precision_x, int precision_y)
...
+	int precision_inv;
```
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:214]: please change int to int32_t

File: src/audio/drc/drc_math_generic.c (hunk 2 above)
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:213]: Change int e to int_32_t e

File: src/audio/drc/drc_math_generic.c [hunk 4]
```diff
@@ -229,11 +234,11 @@ inline int32_t drc_inv_fixed(int32_t x, int precision_x, int precision_y)
...
-		return Q_SHIFT_RND(inv, e, precision_y);
-	if (e < precision_y)
-		return Q_SHIFT_LEFT(inv, e, precision_y);
+	precision_inv = e + qc;
+	if (precision_inv > precision_y)
+		return Q_SHIFT_RND(inv, precision_inv, precision_y);
+	if (precision_inv < precision_y)
+		return Q_SHIFT_LEFT(inv, precision_inv, precision_y);
 	return inv;
 #undef qc
```
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:243]: please change qc to QC

Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:243]: please change qc to QC

**Review by ShriramShastry — COMMENTED** [review]
Changes looks fine

**Review by lgirdwood — COMMENTED** [review]
Will get this into v1.7-rc2

File: src/audio/drc/drc_math_generic.c (hunk 4 above)
Comment (cujomalainey) [reply src/audio/drc/drc_math_generic.c:243]: this is outside the scope of the review, would need another commit/PR

Comment (ShriramShastry) [reply src/audio/drc/drc_math_generic.c:243]: Ok, then I think current modification is good enough to address inverse bug

**Comment by johnylin76** [comment]
Split to 3 commits, PTAL thanks.

File: src/audio/drc/drc_math_generic.c (hunk 2 above)
Comment (lgirdwood) [reply src/audio/drc/drc_math_generic.c:213]: ack, if these should be the same types to make the algo portable to non xtensa arch.

File: src/audio/drc/drc_math_generic.c [hunk 5]
```diff
@@ -183,6 +183,10 @@ inline int32_t drc_asin_fixed(int32_t x)
  */
 inline int32_t drc_pow_fixed(int32_t x, int32_t y)
 {
+	/* Negative input x is not supported. */
+	if (x <= 0)
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_generic.c:187]: comment does not match the condition (as zero is also not allowed)

File: src/audio/drc/drc_math_generic.c [hunk 6]
```diff
@@ -177,7 +177,7 @@ inline int32_t drc_asin_fixed(int32_t x)
 }
 
 /*
- * Input x is Q6.26: (-32.0, 32.0)
+ * Input x is Q6.26; valid range: [0.0, 32.0); negative values are not supported
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_generic.c:180]: This does not the condition in the previous patch, I assume here we support >=0 ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]

File: src/audio/drc/drc_math_generic.c (hunk 2 above)
Comment (cujomalainey) [reply src/audio/drc/drc_math_generic.c:213]: @johnylin76 this change is still missing

File: src/audio/drc/drc_math_generic.c (hunk 3 above)
Comment (cujomalainey) [reply src/audio/drc/drc_math_generic.c:214]: ditto here

**Comment by cujomalainey** [comment]
> make it 1 to 32. Unsupported cases log(0)/log10(0)/log2(0) = -Inf

yep, range is now non inclusive of 0

File: src/audio/drc/drc_math_generic.c [hunk 7]
```diff
@@ -177,7 +177,7 @@ inline int32_t drc_asin_fixed(int32_t x)
 }
 
 /*
- * Input x is Q6.26: (-32.0, 32.0)
+ * Input x is Q6.26; valid range: (0.0, 32.0); x <= 0 is not supported
```
Comment (lgirdwood) [inline src/audio/drc/drc_math_generic.c:180]: Sorry @ShriramShastry @cujomalainey I'm not following your previous comments, do we mean the range should be 1..31 here ? It just looks like the comment could be wrong ?

Comment (ShriramShastry) [reply src/audio/drc/drc_math_generic.c:180]: Yes, Liam. We don't support 0 value while computing the logarithm. Johnny's fix will return back the function without processing the logarithm function if input value is 0. 
Comment section change was requested only to indicate valid supported range is  [1.0 to 32.0].

Comment (cujomalainey) [reply src/audio/drc/drc_math_generic.c:180]: Maybe it's a regional notation? But at least in NA ( represents non-inclusive range, where as [  an inclusive range.

So `(0, 32]` means 0 to 32 not including 0 but including 32. So 0.1 is in the rang with this notation.

**Comment by cujomalainey** [comment]
@johnylin76 i think we are just missing the type fixes then we are good

**Comment by johnylin76** [comment]
> @johnylin76 i think we are just missing the type fixes then we are good

Finished the type fixes. Thanks a lot.

File: src/audio/drc/drc_math_generic.c [hunk 8]
```diff
@@ -210,8 +210,8 @@ inline int32_t drc_inv_fixed(int32_t x, int precision_x, int precision_y)
 	const int32_t A2 = Q_CONVERT_FLOAT(33.57208251953125f, qc);
 	const int32_t A1 = Q_CONVERT_FLOAT(-21.25031280517578125f, qc);
 	const int32_t A0 = Q_CONVERT_FLOAT(7.152250766754150390625f, qc);
-	int e;
-	int precision_inv;
+	int32_t e;
+	int32_t precision_inv;
 	int sqrt2_extracted = 0;
```
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:215]: can we do int to int32_t ?

**Review by ShriramShastry — COMMENTED** [review]
Changes looks fine

File: src/audio/drc/drc_math_generic.c [hunk 9]
```diff
@@ -195,7 +195,7 @@ inline int32_t drc_pow_fixed(int32_t x, int32_t y)
  * Input depends on precision_x
  * Output depends on precision_y
  */
-inline int32_t drc_inv_fixed(int32_t x, int precision_x, int precision_y)
+inline int32_t drc_inv_fixed(int32_t x, int32_t precision_x, int32_t precision_y)
 {
 #define qc 25
```
Comment (ShriramShastry) [inline src/audio/drc/drc_math_generic.c:200]: Can qc be defined in capital [QC] instead of small?

**Review by ShriramShastry — COMMENTED** [review]
Changes looks fine to me

**Comment by cujomalainey** [comment]
> Changes looks fine to me

@ShriramShastry if changes are good please mark "approved" when doing a code review.

**Review by cujomalainey — APPROVED** [review]

**Comment by cujomalainey** [comment]
Also, @ShriramShastry , we should look eventually at moving common math code to the math lib and de-duplicating functions used across multiple audio components

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
CI unrelated.
